core/test/mpi/distributed/vector.cpp

```diff
@@ -0,0 +1,864 @@
+            vec->get_local(),
+            I<I<value_type>>(
+                {{0, 1}, {2, 3}, {4, 5}, {6, 7}, {8, 9}, {10, 11}}),
+            r<value_type>::value);
+    } else {
+        GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(),
+                                    gko::dim<2>(0, 2));
+    }
+}
+
+
+TYPED_TEST(VectorCreation, CanCreateFromLocalVectorAndSize)
+{
+    using dist_vec_type = typename TestFixture::dist_vec_type;
+    using dense_type = typename TestFixture::dense_type;
+    auto local_vec = dense_type::create(this->ref);
+    local_vec->read(this->md_localized[this->comm.rank()]);
+    auto clone_local_vec = gko::clone(local_vec);
+
+    auto vec = dist_vec_type::create(this->ref, this->comm, gko::dim<2>{6, 2},
+                                     local_vec.get());
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec, gko::dim<2>(6, 2));
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), clone_local_vec, 0);
+}
+
+
+TYPED_TEST(VectorCreation, CanCreateFromLocalVectorWithoutSize)
+{
+    using dist_vec_type = typename TestFixture::dist_vec_type;
+    using dense_type = typename TestFixture::dense_type;
+    auto local_vec = dense_type::create(this->ref);
+    local_vec->read(this->md_localized[this->comm.rank()]);
+    auto clone_local_vec = gko::clone(local_vec);
+
+    auto vec = dist_vec_type::create(this->ref, this->comm, local_vec.get());
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec, gko::dim<2>(6, 2));
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), clone_local_vec, 0);
+}
+
+
+template <typename ValueType>
+class VectorReductions : public ::testing::Test {
+public:
+    using value_type = ValueType;
+    using local_index_type = gko::int32;
+    using global_index_type = gko::int64;
+    using part_type =
+        gko::distributed::Partition<local_index_type, global_index_type>;
+    using md_type = gko::matrix_data<value_type, global_index_type>;
+    using dist_vec_type = gko::distributed::Vector<value_type>;
+    using dense_type = gko::matrix::Dense<value_type>;
+
+    VectorReductions()
+        : ref(gko::ReferenceExecutor::create()),
+          comm(MPI_COMM_WORLD),
+          part(gko::share(part_type::build_from_contiguous(
+              this->ref, {ref, {0, 2, 4, 6}}))),
+          vec_a(dist_vec_type::create(ref, comm)),
+          vec_b(dist_vec_type::create(ref, comm))
+    {
+        md_type md_a{{0, 1}, {2, 3}, {4, 5}, {6, 7}, {8, 9}, {10, 11}};
+        md_type md_b{{10, -11}, {8, -9}, {-6, 7}, {4, -5}, {2, -3}, {0, 1}};
+
+        vec_a->read_distributed(md_a, part.get());
+        vec_b->read_distributed(md_b, part.get());
+    }
+
+    void SetUp() override { ASSERT_GE(this->comm.size(), 3); }
+
+    std::shared_ptr<gko::Executor> ref;
+    gko::mpi::communicator comm;
+    std::shared_ptr<part_type> part;
+
+    std::unique_ptr<dist_vec_type> vec_a;
+    std::unique_ptr<dist_vec_type> vec_b;
+};
+
+
+TYPED_TEST_SUITE(VectorReductions, gko::test::ValueTypes);
+
+
+TYPED_TEST(VectorReductions, ComputesDotProduct)
```

Comment:
I think these tests are not belong to core test

Reply:
This question of where to place tests has a bit of a long-winded answer. Originally, everything in `core` tested only code that called no kernels anywhere. Then this became infeasible because even `read` became a kernel call, so now we need to build `reference` kernels every time. In total, I think this would be best suited for `test`, because it also involves the interaction with device memory for different executors.

putting them in `reference` is not an issue, since all reference tests are linked against `ginkgo`, not `ginkgo_reference` alone.

Reply:
I think putting it into test is good because the test also need checked on gpu with gpu-aware mpi